artman/cli/configure.py

```diff
@@ -46,107 +51,116 @@ def configure(log_level=logging.INFO):
-    logger.info('Where do you want to publish code by default?')
-    logger.info('The common valid options are "github" and "local".')
-    publish = six.moves.input('Default publisher: ').lower()
-    try:
-        importlib.import_module('artman.tasks.publish.%s' % publish)
-        return publish
-    except ImportError:
-        logger.error('Invalid publisher.')
-        return _configure_publish()
-
-
-def _configure_github(github):
-    """Determine and return the GitHub configuration.
-
-    Args:
-        github (dict): The current GitHub configuration.
+def _configure_github_config():
+    """Determine and return artman user GitHub config.
 
     Returns:
         dict: The new GitHub configuration.
     """
-    answer = copy(github)
+    answer = GitHubConfig()
     logger.info('Since you intend to publish to GitHub, you need to '
                 'supply credentials.')
     logger.info('Create an access token at: '
                 'https://github.com/settings/tokens')
     logger.info('It needs the "repo" scope and nothing else.')
-    while not answer.get('username'):
-        answer['username'] = six.moves.input('GitHub username: ')
-    while not answer.get('token'):
-        answer['token'] = getpass.getpass('GitHub token (input is hidden): ')
+    while not answer.username:
+        answer.username = six.moves.input('GitHub username: ')
+    while not answer.token:
+        answer.token = getpass.getpass('GitHub token (input is hidden): ')
+
     return answer
+
+def _convert_json(d):
+    """Convert the dict to turn all key into snake case."""
+    new_d = {}
+    for k, v in d.items():
+        if isinstance(v, dict):
+            new_d[stringcase.snakecase(k)] = _convert_json(v)
+        elif isinstance(v, list):
+            if isinstance(v[0], dict):
+                result = []
+                for d2 in v:
+                    result.append(_convert_json(d2))
+                new_d[stringcase.snakecase(k)] = result
+            else:
+                new_d[stringcase.snakecase(k)] = v
+        else:
+            new_d[stringcase.snakecase(k)] = v
+    return new_d
+
+
+def _write_pb_to_yaml(pb, output):
+    # Add yaml representer so that yaml dump can dump OrderedDict. The code
+    # is coming from https://stackoverflow.com/questions/16782112.
+    yaml.add_representer(OrderedDict, _represent_ordereddict)
+
+    json_obj = _order_dict(_convert_json(json.loads(MessageToJson(pb))))
+    with open(output, 'w') as outfile:
+        yaml.dump(json_obj, outfile, default_flow_style=False)
+
+
+def _represent_ordereddict(dumper, data):
+    value = []
+    for item_key, item_value in data.items():
+        node_key = dumper.represent_data(item_key)
+        node_value = dumper.represent_data(item_value)
+
+        value.append((node_key, node_value))
+
+    return yaml.nodes.MappingNode(u'tag:yaml.org,2002:map', value)
+
+
+def _order_dict(od):
+    # The whole key order is flattened which is okay for artman user config
```

Comment:
> The whole key order is flattened

What does this mean?

Reply:
Say an ideal artman user config is in the following field order

"""
field1:
  field1_a:
  field1_b:
  field_c:

field2:
  field2_a:
  field2_b:
  field_c:
"""

The order is defined as "field1, field1_a, field1_b, field2, field2_a, field2_b, field_c" in a flattend manner (vs nested). Note: there is the limitation to its expression as when `field_c` (field name) is defined inside multiple parent fields, it might not end up working. But this won't be a problem in practice.

Reply:
If you intend users to define the data in a flattened manner, why not define the proto in a flattened manner?

Reply:
While artman only has 3 fields, it is reasonable to have everything flattened without impacting understand-ability. For the long run (if any), it will be more flexible using a structured configuration. That is my personal preference though.

Reply:
Either way is fine, but I do not understand why there would be a discrepancy between the proto and the YAML.

Reply:
I did mean to make these fields consistent between proto definition and yaml output. WIthout this ordering implementation, after artman user run `configure-artman` it will end up with the following:

"""
github:
  token: token
  username: test
local:
  toolkit: /toolkit
"""

I would prefer `local` be in front of `github` though it is okay to change. It would be awkward to have `token` in front of `username`.

Reply:
Sorry, I don't follow. What does ordering have to do with flattened vs. nested?

Reply:
It is confusing to call it flattening. Updated the comment.

Reply:
Thanks, I think I understand now: what is flattened is the ordering relation you've defined, not the data structure.